src/Razor/src/Microsoft.VisualStudio.RazorExtension/SyntaxVisualizer/IntraTextAdornmentTagger.cs

```diff
@@ -0,0 +1,247 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.VisualStudio.RazorExtension.SyntaxVisualizer
+{
+    /// <summary>
+    /// Helper class for interspersing adornments into text.
+    /// </summary>
+    /// <remarks>
+    /// To avoid an issue around intra-text adornment support and its interaction with text buffer changes,
+    /// this tagger reacts to text and color tag changes with a delay. It waits to send out its own TagsChanged
+    /// event until the WPF Dispatcher is running again and it takes care to report adornments
+    /// that are consistent with the latest sent TagsChanged event by storing that particular snapshot
+    /// and using it to query for the data tags.
+    /// </remarks>
+    internal abstract class IntraTextAdornmentTagger<TData, TAdornment>
```

Comment:
This file is copied wholesale from the VS SDK sample for a tagger. Don't ask me what it does 😛